### PR TITLE
Ksagiyam/hex interior facet

### PR DESCRIPTION
**Pull request by ksagiyam**
Enable interior facet integration on hex by applying correct permutation to the quadrature points on each side (`+` and `-`).

Note:
Another way of handling interior facet integration on hex would be to choose reference cell - physical cell mapping on each side so that permutation would not be needed.

Depend on:
https://github.com/firedrakeproject/tsfc/pull/317
https://github.com/FInAT/FInAT/pull/138
https://github.com/firedrakeproject/fiat/pull/83

### CODE REVIEW

**Comment by ksagiyam** [comment]
docs build fails as there are TSFC changes. Docs build fine locally.

**Review by dham — DISMISSED** [review]

File: firedrake/mesh.py [hunk 1]
```diff
@@ -317,6 +319,34 @@ def facet_cell_map(self):
...
+            self.mesh.entity_orientations[:, offsets[-3]:offsets[-2]][self.facet_cell[:shape[0], :]],
+            local_facets.reshape(shape + (1, )),
+            axis=2,
+        ).reshape(shape)
+        return op2.Dat(
+            self.local_facet_dat.dataset,
+            data,
+            dtype,
+            "%s_%s_local_facet_orientation" %
+            (self.mesh.name, self.kind),
```
Comment (connorjward) [inline firedrake/mesh.py:347]:
```suggestion
            f"{self.mesh.name}_{self.kind}_local_facet_orientation"
```

File: firedrake/mesh.py [hunk 2]
```diff
@@ -317,6 +319,34 @@ def facet_cell_map(self):
...
+        data = np.empty_like(self.local_facet_dat.data_ro_with_halos).reshape((-1, self._rank))
+        data.fill(np.iinfo(dtype).max)
+        local_facets = self.local_facet_dat.data_ro
+        local_facets = local_facets.reshape((-1, self._rank))
+        shape = local_facets.shape
+        data[:shape[0], :] = np.take_along_axis(
+            self.mesh.entity_orientations[:, offsets[-3]:offsets[-2]][self.facet_cell[:shape[0], :]],
+            local_facets.reshape(shape + (1, )),
+            axis=2,
+        ).reshape(shape)
```
Comment (connorjward) [inline firedrake/mesh.py:341]: This is really confusing. You `reshape` multiple times and also define a `shape` variable etc. Also the slicing that you do is difficult to understand. Could this be refactored?

File: tests/regression/test_integral_hex.py [hunk 3]
```diff
@@ -18,3 +18,38 @@ def test_integral_hex_exterior_facet(mesh_from_file, family):
...
+    f = Function(V).interpolate(2 * x + 3 * y + 5 * z)
+    u = TrialFunction(V)
+    v = TestFunction(V)
+    a = inner(u('+'), v('+')) * dS(degree=3)
+    L = inner(f('+'), v('-')) * dS(degree=3)
+    bc = DirichletBC(V, f, "on_boundary")
+    sol = Function(V)
+    solve(a == L, sol, bcs=[bc])
+    err = assemble((sol - f)**2 * dx)
+    assert err < 1.e-28
```
Comment (dham) [inline tests/regression/test_integral_hex.py:55]:
```suggestion
    err = assemble((sol - f)**2 * dx)**0.5
    assert err < 1.e-14
```

File: tests/regression/test_integral_hex.py [hunk 4]
```diff
@@ -18,3 +18,38 @@ def test_integral_hex_exterior_facet(mesh_from_file, family):
...
+@pytest.mark.parametrize('family', ["Q", "DQ"])
+def test_integral_hex_interior_facet(mesh_from_file, family):
+    if mesh_from_file:
+        mesh = Mesh(join(cwd, "..", "meshes", "cube_hex.msh"))
+    else:
+        mesh = UnitCubeMesh(2, 3, 5, hexahedral=True)
+    V = FunctionSpace(mesh, family, 3)
+    x, y, z = SpatialCoordinate(mesh)
+    f = Function(V).interpolate(2 * x + 3 * y * y + 4 * z * z * z)
+    assert assemble((f('+') - f('-'))**2 * dS) < 1.e-28
```
Comment (dham) [inline tests/regression/test_integral_hex.py:34]:
```suggestion
    assert assemble((f('+') - f('-'))**2 * dS)**0.5 < 1.e-14
```

**Review by connorjward — CHANGES_REQUESTED** [review]

File: firedrake/mesh.py (hunk 2 above)
Comment (ksagiyam) [reply firedrake/mesh.py]: Fixed. Please check if this is now fine.

Comment (connorjward) [reply firedrake/mesh.py]: Yeah that's much better thanks.

**Comment by ksagiyam** [comment]
Reviewed in a Firedrake meeting. Docs built fine locally.

**Pull request merged**